### PR TITLE
nrf_802154_sl: Introduce nRF IEEE 802.15.4 Service Layer

### PR DESCRIPTION
**Pull request by jciupis**
This commit introduces nRF IEEE 802.15.4 Service Layer library required
for advanced features of the nRF IEEE 802.15.4 radio driver.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

### CODE REVIEW

File: Kconfig.nrfxlib [hunk 1]
```diff
@@ -16,4 +16,5 @@ rsource "nrf_rpc/Kconfig"
 rsource "zboss/Kconfig"
 rsource "openthread/Kconfig"
```
Comment (rlubos) [inline Kconfig.nrfxlib:18]: This empty line should rather be below `rsource "nrf_802154_sl/Kconfig"` I believe.

File: nrf_802154_sl/include/nrf_802154_sl_stats.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define nrf_802154_sl_stat_counter_increment(...)
```
Comment (rlubos) [inline nrf_802154_sl/include/nrf_802154_sl_stats.h:39]: No include guard?

File: nrf_802154_sl/include/fem/nrf_fem_protocol_api.h [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+#if defined(NRF52_SERIES)
+#include "hal/nrf_ppi.h"
+#elif defined(NRF53_SERIES)
+#include "hal/nrf_dppi.h"
+#else
+#error Unsupported chip
+#endif
+
+typedef enum
```
Comment (rlubos) [inline nrf_802154_sl/include/fem/nrf_fem_protocol_api.h:66]: This and several other files are missing `extern "C"` block, probably should be fixed for future releases.

**Comment by hurricaneFromMoscow** [comment]
Did you add all these API headers which looks like a copy of MPSL ones intentionally?

**Comment by jciupis** [comment]
@hurricaneFromMoscow assuming you're asking about the PA/LNA headers - yes, I added them intentionally. Although available in MPSL, these headers are not yet used in the 802.15.4 driver. They will be removed when the driver is ported to use MPSL API.

**Review by rlubos — APPROVED** [review]

**Review by czeslawmakarski — APPROVED** [review]

**Review by maciekfabia — APPROVED** [review]

**Review by thst-nordic — APPROVED** [review]

**Review by pkwiek — APPROVED** [review]

File: nrf_802154_sl/include/fem/nrf_fem_protocol_api.h [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+ *
+ * @retval   ::NRFX_SUCCESS               PA activate setup purge is successful.
+ * @retval   ::NRFX_ERROR_FORBIDDEN       PA is currently disabled.
+ * @retval   ::NRFX_ERROR_INVALID_STATE   PA activate setup purge could not be performed due to invalid or missing configuration parameters,
+ *                                        that were previously used in @ref nrf_802154_fal_pa_configuration_set call.
+ */
+int32_t nrf_802154_fal_pa_configuration_clear(void);
+
+/**
+ * @brief Sets up LNA using the provided event for the upcoming radio reception.
```
Comment (b-gent) [inline nrf_802154_sl/include/fem/nrf_fem_protocol_api.h:158]: This is too much for doxygen doc in a header file, should be moved to RST documentation in the future. Make sure to follow up with @FrancescoSer

File: nrf_802154_sl/include/fem/nrf_fem_protocol_api.h [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            uint32_t register_address;        /* Address of event register. */
+        } generic;
+    }                           event;
+    bool                        override_ppi; /* False to ignore the PPI channel below and use the one set by application. True to use the PPI channel below. */
+    uint8_t                     ppi_ch_id;    /* PPI channel to be used for this event. */
+    nrf_802154_fal_event_type_t type;         /* Type of event source. */
+} nrf_802154_fal_event_t;
+
+/**
+ * @brief Sets up PA using the provided events for the upcoming radio transmission.
```
Comment (b-gent) [inline nrf_802154_sl/include/fem/nrf_fem_protocol_api.h:115]: This is too much for doxygen doc in a header file, should be moved to RST documentation in the future. Make sure to follow up with @FrancescoSer

File: nrf_802154_sl/include/fem/simple_gpio/nrf_fem_config.h [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ */
+
+/** Time in microseconds when PA GPIO is activated before the radio is ready for transmission. */
+#define NRF_FEM_SIMPLE_GPIO_PA_TIME_IN_ADVANCE_US  23
+
+/** Time in microseconds when LNA GPIO is activated before the radio is ready for reception. */
+#define NRF_FEM_SIMPLE_GPIO_LNA_TIME_IN_ADVANCE_US 5
+
+/**
+ * @section Gains.
```
Comment (b-gent) [inline nrf_802154_sl/include/fem/simple_gpio/nrf_fem_config.h:117]: This happens in a couple places, ``@section`` is something you don't want to use in a header file,  You probably want ``@name``

**Comment by b-gent** [comment]
My comments are not blocking, they are remarks for future enhancement of the doc that will be driven by @FrancescoSer

**Review by b-gent — APPROVED** [review]

**Review by FrancescoSer — APPROVED** [review]
Approved, but we have to expand the documentation in the future.

**Comment by ioannisg** [comment]
@jciupis can you look atht e CI and checkpatch errors?

**Comment by thst-nordic** [comment]
checkpatch was disabled, the failure is from an old run of the CI.  this can merge.

**Pull request merged**